internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,755 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// agentServiceConfigName is the one and only name for an AgentServiceConfig
+	// supported in the cluster. Any others will be ignored.
+	agentServiceConfigName = "agent"
+	// filesystemPVCName is the name of the PVC created for assisted-service's filesystem.
+	filesystemPVCName = "agent-filesystem"
+	// databasePVCName is the name of the PVC created for postgresql.
+	databasePVCName = "agent-database"
+
+	name                                = "assisted-service"
+	databaseName                        = "postgres"
+	databaseSecretName                  = databaseName
+	databasePort                  int32 = 5432
+	servicePort                   int32 = 8090
+	assistedServiceDeploymentName       = "assisted-service"
+	postgresDeploymentName              = databaseName
+
+	// assistedServiceContainerName is the Name property of the assisted-service container
+	assistedServiceContainerName string = "assisted-service"
+	// databaseContainerName is the Name property of the postgres container
+	databaseContainerName string = databaseName
+)
+
+// AgentServiceConfigReconciler reconciles a AgentServiceConfig object
+type AgentServiceConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	// the Namespace the operator is running in
+	// TODO(djzager): This must be configured on the operator deployment (ie. in the CSV
+	// as from the metadata.Namespace
+	Namespace string
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=route.openshift.io,resources=routes,verbs=get;list;watch;create;update;patch;delete
+
+func (r *AgentServiceConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	_ = r.Log.WithValues("agentserviceconfig", req.NamespacedName)
+
+	instance := &adiiov1alpha1.AgentServiceConfig{}
+
+	// We only support one AgentServiceConfig per cluster, and it must be called "hive". This prevents installing
+	// AgentService more than once in the cluster.
+	if req.NamespacedName.Name != agentServiceConfigName {
+		err := fmt.Errorf("Invalid name (%s) expected %s", req.NamespacedName.Name, agentServiceConfigName)
```

Comment:
weeds: logging at the "error" level is best reserved for application errors where the admin of the workload should take action to fix something. Could not connect to a DB, ran out of memory, recovered from some bug, etc.

Bad user input is not worth setting off "error alarms" in the log. I'd stick to Info.